music_service_async_interface/__init__.py

```diff
@@ -215,41 +317,162 @@ async def get_file_url(
+        if ommited value from session is used
+        :param kwargs: allows subfunction to define custom keyword arguments
+        :raises InsufficientAudioQuality: when available :class:`AudioQuality` is lower than `required_quality`
+        :return: direct URL of music file
         """
         ...
 
-    if AsyncSeekableHTTPFile is not None:
-
-        async def get_async_file(
-            self,
-            required_quality: Optional[AudioQuality] = None,
-            preferred_quality: Optional[AudioQuality] = None,
-            filename: Optional[str] = None,
-            **kwargs,
-        ) -> AsyncSeekableHTTPFile:
-            """Gets async `filelike` object for Track
-
-            :param required_quality: Quality requirement for track
-            :param preferred_quality: Preferred audio quality, which we are asking for
-            :param filename: File name for `filelike`
-            :raise: :class:`InsufficientAudioQuality` when quality requirements are not met
-            :return: Music file
-            """
+    async def get_async_file(
+        self,
+        required_quality: Optional[AudioQuality] = None,
+        preferred_quality: Optional[AudioQuality] = None,
+        filename: Optional[str] = None,
+        **kwargs,
+    ) -> AsyncSeekableHTTPFile:
+        """Gets async `filelike` object containing track file
+
+        :param required_quality: required (minimal accepted) :class:`AudioQuality` for track
+        if ommited value from session is used
+        :param preferred_quality: preferred (maximum) :class:`AudioQuality` you want to get
+        if ommited value from session is used
+        :param filename: filename for `filelike` object
+        :param kwargs: allows subfunction to define custom keyword arguments
+        :raises ImportError: when `httpseekablefile` library is not installed
+        :raises InsufficientAudioQuality: when available :class:`AudioQuality` is lower than `required_quality`
+        :return: async `filelike` with music file
+        """
+        if AsyncSeekableHTTPFile is None:
+            raise ImportError("httpseekablefile not installed")
 
-            return await AsyncSeekableHTTPFile.create(
-                await self.get_file_url(required_quality, preferred_quality, **kwargs),
-                self.title if filename is None else filename,
-                self.sess.sess,
-            )
+        return await AsyncSeekableHTTPFile.create(
+            await self.get_file_url(required_quality, preferred_quality, **kwargs),
+            self.title if filename is None else filename,
+            self.sess.sess,
+        )
 
 
 class ObjectCollection:
+    """Base class for any collections of objects
+
+    Makes sure particular method for object type is overridden.
+    Allows easy iteration over all object types in collection.
+
+    wrong example:
+    >>> class Album(ObjectCollection[Track]):
```

Comment:
I'm not sure we should include anti-patterns and wrong examples. People often copy the examples blindly and just hope for them to work. Also it's easier to understand when there are working examples everywhere. We could simply say that the `ObjectCollection[Track]` class has to have the `tracks()` method.

Reply:
explanation below

Reply:
k, we met some consensus